adr/00-packages.md

```diff
@@ -0,0 +1,105 @@
+# Architecture Decision Record: Package structure
+
+## Context
+
+We need to specify structure for the application. 
+It will involve choosing how we will divide code into packages and modules. 
+We would like to make a clear decision, what application structure should be going forward.
+
+## Decision
+
+We will divide application by functionalities into packages.
+
+We decided on the following packages:
+* common
+* auth
+* articles
+* comments
+* users
+
+The common package will contain auxiliary structures that will be used in the project.
+The auth package will contain the tools needed to authorize the user.
+
+Packages of articles, comments and users will have a similar structure. 
+Each of them will contain specific endpoints, will service them using services and will communicate with the database using repositories.
+In these packages will be clearly separated api layer that will contain "the contract", i.e. the endpoint descriptions, without logic.
+  
+Example with one implementation per service:
+
+```
+common
+├── BaseEndpoints
+├── Configuration
+├── Pagination
+├── ErrorInfo
+└── Exceptions
+auth
+└── AuthService
+articles
+├── api
+│   ├── ArticleResponse
+│   ├── ArticlesListResponse
+│   ├── ArticleCreateData
+│   ├── ArticleCreateRequest
+│   ├── ArticleUpdateData
+│   ├── ArticleUpdateRequest
+│   └── ArticlesEndpoints
+├── Author
+├── Article
+├── ArticlesRepository
+├── ArticlesService
+├── Tag
+├── TagsRepository
+├── TagsService
+└── ArticlesServerEndpoints
+comments
+├── api
+│   ├── CommentResponse
+│   ├── CommentsListResponse
+│   ├── CommentCreateData
+│   ├── CommentCreateRequest
+│   ├── CommentUpdateData
+│   ├── CommentUpdateRequest
+│   └── CommentsEndpoints
+├── Author
+├── Comment
+├── CommentsRepository
+├── CommentsService
+└── CommentsServerEndpoints
+users
+├── api
+│   ├── ProfileResponse
+│   ├── UserResponse
+│   ├── UserRegisterData
+│   ├── UserRegisterRequest
+│   ├── UserLoginData
+│   ├── UserLoginRequest
+│   ├── UserUpdateData
+│   ├── UserUpdateRequest
+│   └── UsersEndpoints
+├── Profile
+├── User
+├── UsersRepository
+├── UsersService
+└── UsersServerEndpoints
+```
+
+
+## Rationale
+
+* Dividing the code by functionality will help organize it better.
+* Separate pure description of the api can better showcase tapirs strengths. For example, it can be used to generate client for the api.
+* Pure api description, can be easily put into a separate module. This allows other projects to use it as a dependency.
+* Pure api description allow browse pure endpoint declarations as a documentation of the API.
+
+## Consequences
+
+* Maintaining such a structure requires discipline.
+* Conforming to new structure will require refactoring the application.
+* New structure requires more work upfront, when writing new code.
```

Comment:
hm in which areas? endpoints + server endpoints or sth else as well? would be good to be more specific

Reply:
This comment left after the first version of the app, which was more "layered". I think that it can be removed now, also "Maintaining such a structure requires discipline.". There is not as much difference between the new and the old structure as we thought at the beginning, so i don't think now that the new structure requires that much discipline and more work upfront. And what do you think?